Refactor Guest `createAnnotation` method and add missing tests

robertknight opened the pull request:
I noticed it was quite hard to follow the async flow of the `Guest#createAnnotation` method and see which properties of the annotation are guaranteed to be populated before it is sent to the sidebar. Also the method is inherently async but returns synchronously, so there is no way to properly wait for it to complete. This PR refactors `createAnnotation` using async/await to fix that. It also adds some missing tests for this important method.

Changes in detail:

- Refactor `createAnnotation` method of `Guest` to use async/await, which simplifies the control flow. I also renamed a misleading `selectors` variable
- Add tests for triggering this method by clicking on the "Annotate" or "Highlight" buttons in the adder toolbar, which is how this method gets called in the actual application
- Add tests for the `target` field of new annotations, which contains the selectors for the annotation

robertknight commented on src/annotator/guest.js:
```diff
@@ -122,13 +122,13 @@ export default class Guest extends Delegator {
-      onAnnotate: () => {
-        this.createAnnotation();
+      onAnnotate: async () => {
+        await this.createAnnotation();
```
This change allows the tests to check when the activity triggered by clicking "Annotate" or "Highlight" is finished.

robertknight commented on src/annotator/guest.js:
```diff
@@ -535,53 +535,46 @@ export default class Guest extends Delegator {
    *   new annotation, other than `target`, `document` and `uri` which are
    *   set by this method.
+   * @return {Promise<AnnotationData>} - The updated annotation with selectors
+   *   and metadata populated.
```
This method used to return an `AnnotationData` despite being async internally. It now returns a Promise as it should.

robertknight commented on src/annotator/test/guest-test.js:
```diff
@@ -29,10 +29,6 @@ class FakeTextRange {
-// A little helper which returns a promise that resolves after a timeout
-const timeoutPromise = (millis = 0) =>
-  new Promise(resolve => setTimeout(resolve, millis));
-
```
This is no longer needed because tests can explicitly await `createAnnotation`'s results.

LMS007 commented on src/annotator/test/guest-test.js:
```diff
@@ -609,29 +627,55 @@ describe('Guest', () => {
-        assert.equal(annotation.uri, fakeDocumentMeta.uri());
-        assert.deepEqual(annotation.document, fakeDocumentMeta.metadata);
-      });
+      assert.equal(annotation.uri, fakeDocumentMeta.uri());
```
This test, while valid, is testing the mutated param, but not an actual returned variable. I think the function should either return nothing so it's very clear its mutating the param, or it should test the thing it actually is returning.

`const annotation = await guest.createAnnotation({});`

LMS007 commented on src/annotator/guest.js:
```diff
@@ -535,53 +535,46 @@ export default class Guest extends Delegator {
-  createAnnotation(annotation = {}) {
-    const root = this.element;
-
+  async createAnnotation(annotation = {}) {
```
It might be a little more clear to create a new var that gets returned rather than returning the mutated param? 

```
async createAnnotation(partialAnnotation = {}) {
   const annotation = {...partialAnnotation}
```

LMS007 commented on src/annotator/guest.js:
```diff
@@ -535,53 +535,46 @@ export default class Guest extends Delegator {
 
-    const selectors = Promise.all(ranges.map(getSelectors));
-    const targets = Promise.all([info, selectors]).then(setTargets);
+      // In the Hypothesis API the field containing the selectors is called
```
👍

LMS007 reviewed (approved):
`createAnnotation` is less confusing and better tested. My only hangup was the questionable mutation of the `annotation` param and if that is still something we can to continue doing? Otherwise looks good to go.

robertknight commented on src/annotator/guest.js:
```diff
@@ -535,53 +535,46 @@ export default class Guest extends Delegator {
-        selector,
-      }));
-    };
+    const info = await this.getDocumentInfo();
```
> If you agree, perhaps we can do another clean-up pass in the future to address that?

Yes, I agree this needs further cleanup.

robertknight commented on src/annotator/guest.js:
```diff
@@ -535,53 +535,46 @@ export default class Guest extends Delegator {
-    info.then(setDocumentInfo);
+    const root = this.element;
+    const rangeSelectors = await Promise.all(
+      ranges.map(range => this.anchoring.describe(root, range))
```
Good question! - Currently the behavior is the same as before - no annotation gets created and an error is logged into the browser console. A future improvement may be to show some kind of notice and log error details to Sentry.

robertknight commented:
> My only hangup was the questionable mutation of the annotation param and if that is still something we can to continue doing?

Thanks for the feedback. It turns out the only remaining use of the mutable annotation param was to set the `$highlight` flag when clicking on the "Highlight" button. I've replaced this param with an options object that has a single `highlight` option and added some related missing tests.
